SYS-1971: Improve MARC support with Alma API client

akohler opened the pull request:
Implements [SYS-1971](https://uclalibrary.atlassian.net/browse/SYS-1971).

Goals:
* Hide the complexities.  Users should just be able to get the record they need, work with it via pymarc, and put it back.
* Fix legacy code, The API client was OK (though clumsy), but the `alma_marc` routines were named / focused on bib records, though they work with holdings too (and authority records, a later addition).
* Support record creation and deletion.  Currently we can only get and update existing records.  Record creation and deletion will be needed sometimes.

This PR reworks MARC support to achieve (I hope) all of those goals:
* New base class `AlmaMARCRecord`, which implements all of the code to convert Alma API XML responses to Python objects and back again.
  * This has no public methods, other than the constructor, which takes an optional `api_response` and parses data out of that.
  * Attributes which are allowed to be updated are exposed via `__slots__`, which (as implemented) prevents users of the objects from setting attributes which aren't allowed by the Alma API.
  * Attributes which aren't relevant for sending to Alma are not exposed via `__slots__`, so can't be set on these objects.  As desired, we can expose them via read-only `@property` decorators (on subclasses).
  * The base class has two public properties:
    *  `alma_xml`, which wraps the code to build the XML needed by the Alma APIs.  This is built on request, and expected to be requested only by the `AlmaAPIClient` methods when they send it to the APIs.  It's a property, so should be read-only.
    * `marc_record`, which exposes the pymarc record, either retrieved from Alma or provided by external programs.  This is read/write.
* New subclasses `AuthorityRecord`, `BibRecord` and `HoldingRecord`.  These extend the base class by defining new read/write attributes in their own `__slots__` (which add to the base class's values).  I've exposed a few read-only properties as well, and we can add more as needed.

Record retrieval and updating was shown in my draft PR last month.  That hasn't changed, other than code cleanup.  I've added `get_authority_record()` to the client, mainly to confirm my general approach works on those in addition to bibs and holdings.  I haven't added support for updating authority records, I'll do that in future work.

Testing: Edit and experiment with a variation on this, which create a bib record, adds a holdings record to it, then deletes both (in reverse order, due to dependencies).

There currently are no automated tests.  I intend to add some in future work.
```
import argparse
from alma_api_client import AlmaAPIClient
from pymarc import Field, Indicators, Record, Subfield
from alma_api_client.models.marc_records import BibRecord, HoldingRecord


def _get_arguments() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Process metadata for MAMS ingestion.")
    parser.add_argument(
        "--api_key",
        help="SANDBOX API KEY",
        required=True,
    )
    return parser.parse_args()


def main() -> None:
    args = _get_arguments()
    client = AlmaAPIClient(args.api_key)

    # Create a very simple bib record.
    marc_bib = Record()
    title = Field(
        tag="245",
        indicators=Indicators("0", "0"),
        subfields=[Subfield(code="a", value="alma_api_demo test 245")],
    )
    marc_bib.add_field(title)

    print("---Creating bib record---")
    my_bib = BibRecord()
    my_bib.marc_record = marc_bib
    response = client.create_bib_record(my_bib)
    my_new_bib = BibRecord(response)

    print("---Showing new bib record---")
    print(my_new_bib.record_format)
    print(my_new_bib.suppress_from_publishing)
    print(my_new_bib.bib_id)
    print(my_new_bib.marc_record)

    # Create a very simple holding record.
    marc_hol = Record()
    f852 = Field(
        tag="852",
        indicators=Indicators("7", "_"),
        subfields=[
            Subfield(code="b", value="CLICC"),
            Subfield(code="c", value="cccollege"),
        ],
    )
    marc_hol.add_field(f852)

    print("---Creating holding record---")
    my_hol = HoldingRecord()
    my_hol.marc_record = marc_hol
    response = client.create_holding_record(my_new_bib.bib_id, my_hol)
    my_new_hol = HoldingRecord(response)

    print("---Showing new holding record---")
    print(my_new_hol.suppress_from_publishing)
    print(my_new_hol.holding_id)
    print(my_new_hol.marc_record)

    print("---Deleting new holding record---")
    client.delete_holding_record(my_new_bib.bib_id, my_new_hol.holding_id)
    # Still clumsy, for now
    print(response["api_response"]["status_code"])

    print("---Deleting new bib record---")
    client.delete_bib_record(my_new_bib.bib_id)
    # Still clumsy, for now
    print(response["api_response"]["status_code"])


if __name__ == "__main__":
    main()
```


[SYS-1971]: https://uclalibrary.atlassian.net/browse/SYS-1971?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

akohler committed:
Add AlmaMARCRecord class for MARC and associated Alma data

akohler committed:
Add get/update bib/holdings record methods using AlmaMARCRecord object

akohler committed:
Make legacy get_pymarc_record_from_bib more generic, prepare for deprecation

ztucker4 commented:
So far, this looks great to me. It's really nice to be able to think of records retrieved from Alma solely in Pymarc terms, and to have Alma API complexities dealt with automatically.

The only thing I see that might be a little confusing is that the `alma_xml` and the `marc_record` of an `AlmaMARCRecord` aren't synchronized. I haven't needed to modify any part of an Alma record other than the MARC `<record>`, but I could imagine it happening - e.g. we want to update one of the <suppress_from> values, in addition to the MARC record itself. In a case like this, I think it would be slightly clearer if the `alma_xml` value kept its `<record>` in sync with the `marc_record` value without needing to call `prepare_xml_for_update()`. 

Having said that, I think it might be cumbersome/a performance issue to call `prepare_xml_for_update()` automatically whenever a `marc_record` is updated. So, just renaming the method to something like `sync_xml_with_marc_record()` could be a little clearer for cases when we want to see the final XML before we send it back to Alma.

akohler commented:
@ztucker4 Many thanks for the review and feedback.  I'm glad that this is moving in the right direction so far.

> The only thing I see that might be a little confusing is that the `alma_xml` and the `marc_record` of an `AlmaMARCRecord` aren't synchronized .... [for example, if] we want to update one of the <suppress_from> values, in addition to the MARC record itself. In a case like this, I think it would be slightly clearer if the `alma_xml` value kept its `<record>` in sync with the `marc_record` value without needing to call `prepare_xml_for_update()`.

The calls to `prepare_xml_for_update()` are intended to happen solely within the API client, and only before an update (or later, create) API call is made.  I intend to deprecate the existing `alma_marc.prepare_bib_for_update()`.

> Having said that, I think it might be cumbersome/a performance issue to call `prepare_xml_for_update()` automatically whenever a `marc_record` is updated.

Agreed.  For legacy scripts, they should continue calling the old `prepare_bib_for_update()` before calling a legacy `update` client method.

> So, just renaming the method to something like `sync_xml_with_marc_record()` could be a little clearer for cases when we want to see the final XML before we send it back to Alma.

Good point.  I still need to tackle bib and holdings (and maybe authority) creation, where the XML will also be relevant.  I'm not sure users need to see the XML at any point, other than possible debugging,

My current thinking is to provide sub-classes of `AlmaMARCRecord` for bibs / holdings / auths, with specific data elements associated with each.  Those will be regular class attributes, settable by the user, which then get marshalled into the relevant XML before the API call.  Relevant ones will also be parsed out of the XML when a record is retrieved.

The [Ex Libris bib record doc](https://developers.exlibrisgroup.com/alma/apis/docs/xsd/rest_bib.xsd/?tags=GET) has a lot of fields, almost all of them "output" parameters, which I believe means come with the XML we get, but don't need to be sent back.  That'll take some testing.

Thanks --Andy

akohler committed:
WIP implementing AlmaMARCRecord subclasses

akohler committed:
Clean up XML handling and add new bib methods

akohler committed:
Add new holding methods and clean up code

akohler committed:
Add deprecation warnings to legacy methods

akohler commented:
@ztucker4 This is ready for official review.  Details in the updated PR description.  Please be strict: if there are significant problems, let's catch them now, before updating scripts to use this new interface.
Thanks --Andy

akohler committed:
Bump package version to 0.1.0

akohler committed:
Add comment showing example of attributes after conversion from XML

ztucker4 reviewed (changes requested):
Overall, this is a huge improvement from the previous version of the API client. It's easy to use to get, update, create, and delete records. I do have two concerns:

1. Behavior when a record is not found. If I request an invalid record:
```
invalid_bib = client.get_bib_record("invalid_bib_id")
invalid_holding = client.get_holding_record("invalid_bib_id", "invalid_holding_id")
invalid_authority = client.get_authority_record("invalid_authority_id")
```
I get no immediate indication of any problems. If I check the values of these objects' attributes, I can see that they all contain empty strings/`None`s or raise `AttributeErrors`, but it's not obvious that there was no record found. I think it would be preferable if the `get_` methods in the client checked the status code of the API response and raised some kind of error on non-200s.

2. Python version support. Since `pyproject.toml` specifies Python >=3.9, I tried building the package and importing it in a couple of local `venv`s with different Python versions. In Python 3.11 and 3.12, I was not able to import the package. The error messages I saw are below. I didn't try any versions earlier than 3.11.
I don't think we're actively using anything older than Python 3.13 anywhere that might use this package, so it's probably easiest to just update `pyproject.toml`. 

3.11 error: 
```
File "/Users/zoetucker/venvs/311/lib/python3.11/site-packages/alma_api_client/alma_api_client.py", line 18
    type Data = Union[bytes, dict, list[tuple]]
         ^^^^
SyntaxError: invalid syntax
```
3.12 error:
 ```
File "/Users/zoetucker/venvs/312/lib/python3.12/site-packages/alma_api_client/alma_api_client.py", line 4, in <module>
    from warnings import deprecated
ImportError: cannot import name 'deprecated' from 'warnings' (/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/warnings.py). Did you mean: '_deprecated'?
```

akohler commented:
@ztucker4 Thank you for the excellent review comments, this is extremely helpful!

Improvements to API response handling were (and still are) on my internal roadmap, but you're right that it should be addressed now to some degree.  One major point of moving to these objects is to improve handling of state, and that includes correct handling of invalid states.

The testing against earlier versions is also important, and something I didn't even consider.  The Ex Libris server where we run scripts centrally is still on Python 3.11, and might not be easy to upgrade - though I'm looking more at `uv` as a way to manage environments outside of Docker, and maybe DigInf team will allow that.  Meanwhile though, I'll see about downgrading my Types and Deprecations to 3.11.

Thanks again --Andy

akohler committed:
Drop environment to Python 3.11 for compatibility with our oldest target system

akohler committed:
Raise ValueError when record cannot be retrieved using new methods

akohler commented:
@ztucker4 This is ready for another review.  Two fairly small changes:
1. I dropped back to Python 3.11, as that's our main target environment for now.  I tweaked the type definition (requires 3.12) and added `typing_extensions` which backports other Typing changes like `deprecated` (requires 3.13),
2. I added a status check, narrowly targeted to `_get_marc_record()` (and methods using it) for now.  The proper solution is to rework the API response handling where it happens, probably by adding an `ApiResponse` class which makes several things friendlier, but that's broader scope than I want to do for now as it will require changes to nearly all client methods.

This means the new `get_TYPE_record()` methods can raise `ValueError`, the only built-in exception which seems relevant:
```
>>> client.get_bib_record("INVALID_ID")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/app_user/project/src/alma_api_client/alma_api_client.py", line 513, in get_bib_record
    api_response = self._get_marc_record(api, parameters)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/app_user/project/src/alma_api_client/alma_api_client.py", line 486, in _get_marc_record
    raise ValueError(f"Unable to get MARC record for {api}")
ValueError: Unable to get MARC record for /almaws/v1/bibs/INVALID_ID
```
Scripts using this should try/catch:
```
for bib_id in bib_ids:
  try:
    record = client.get_bib_record(bib_id)
    # Do something
  except ValueError...:
    # log/continue whatever
```

I tested in the 3.11 container, no problems found.  I also installed the package via git branch into a 3.13 container, where it built cleanly, again no problems found.

I checked all package files in VSCode, in the 3.11 container, no warnings or exceptions.  I also ran `pylint` and got hundreds of messages, because it's extremely picky by default (every TODO line, every missing docstring, and more) but nothing of substance found.  My own TODO: Run things through default `pylint` and marvel sadly at how much better I can get, someday.

Please take a fresh look.  Thanks --Andy

ztucker4 reviewed (changes requested):
The package appears to now be compatible with Python 3.11, 3.12, and 3.13. The `ValueError`s make sense and will be useful in the try/except structure you suggest. 

However, I'm currently getting this `ValueError` on all requests for both valid and invalid records. I think this is because the `status_code` value checked in `_get_marc_record()` is an int, not a str. Could you check this?

akohler committed:
Fix type problem when checking response status

akohler commented:
@ztucker4 You're absolutely right, thank you!  Fixed, please confirm:
```
>>> client.get_bib_record("NONE")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/app_user/project/src/alma_api_client/alma_api_client.py", line 513, in get_bib_record
    api_response = self._get_marc_record(api, parameters)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/app_user/project/src/alma_api_client/alma_api_client.py", line 486, in _get_marc_record
    raise ValueError(f"Unable to get MARC record for {api}")
ValueError: Unable to get MARC record for /almaws/v1/bibs/NONE

>>> client.get_bib_record("9992524713606533")
<alma_api_client.models.marc_records.BibRecord object at 0x78a8f709a140>
```

ztucker4 reviewed (approved):
This all looks good to me now! I've tested basic functionality in 3.11, 3.12, and 3.13. I'm able to get, create, update, and delete bib and holding records, and get authority records.

The code itself is clear and is well documented. The classes in `marc_records.py` make sense and are easy to use.

I think this is all in a good state for now. Merging, thanks!
